File: LeetCode/Java Solutions/Maximum Number of Points with Cost/maximum.java
class Solution {
    public long maxPoints(int[][] points) {
        int m = points.length;
        int n = points[0].length;
        long[] dp = new long[n];
        
        // Initialize dp with the first row
        for (int j = 0; j < n; j++) {
            dp[j] = points[0][j];
        }
        
        for (int i = 1; i < m; i++) {
            long[] newDp = new long[n];
            long leftMax = 0;
            
            // Left to right
            for (int j = 0; j < n; j++) {
                leftMax = Math.max(leftMax - 1, dp[j]);
                newDp[j] = leftMax + points[i][j];
            }
            
            long rightMax = 0;
            // Right to left
            for (int j = n - 1; j >= 0; j--) {
                rightMax = Math.max(rightMax - 1, dp[j]);
                newDp[j] = Math.max(newDp[j], rightMax + points[i][j]);
            }
            
            dp = newDp;
        }
        
        long maxPoints = 0;
        for (long value : dp) {  // Changed 'points' to 'value'
            maxPoints = Math.max(maxPoints, value);
        }
        
        return maxPoints;
    }
}

File: LeetCode/Python Solutions/Maximum Number of Points with Cost/maximum.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        m, n = len(points), len(points[0])
        dp = points[0]
        
        for i in range(1, m):
            new_dp = [0] * n
            left_max = 0
            
            # Left to right
            for j in range(n):
                left_max = max(left_max - 1, dp[j])
                new_dp[j] = left_max + points[i][j]
            
            right_max = 0
            # Right to left
            for j in range(n - 1, -1, -1):
                right_max = max(right_max - 1, dp[j])
                new_dp[j] = max(new_dp[j], right_max + points[i][j])
            
            dp = new_dp
        
        return max(dp)